src/replace-positional-args.js

```diff
@@ -0,0 +1,36 @@
+// Loops over script tokens and replace any position args where positional args are denoted by $<number> and are 1
+// indexed (like bash scripts), eg `echo hello $1 world`
+// eslint-disable-next-line complexity
+function replacePositionalArgs(script, args) {
+  if (!script) {
+    return [script, args]
+  }
+  const scriptTokens = script.split(' ')
+  const usedArgs = []
+  for (const [i, token] of scriptTokens.entries()) {
+    if (token[0] !== '$') {
+      // eslint-disable-next-line no-continue
+      continue
+    }
+    const argIndex = parseInt(token.slice(1), 10)
+    const arg = args[argIndex]
+    if (!arg) {
+      throw new Error(`Argument $${argIndex} not specified`)
+    }
+    scriptTokens[i] = args[argIndex]
+    usedArgs.push(argIndex)
+  }
+  const reducedArgs = args.filter((_, i) => !usedArgs.includes(i))
```

Comment:
Rather than this, why not splice `args` above instead of `usedArgs.push(argIndex)`?

Reply:
I didn't want to operate on the function input, though i suppose i could make a copy. The bigger issue though is that doesn't work if a positional arg is used more than once. I added a test to demonstrate this, but the following needs to work to: `foobar $1 $2 $1`

If you splice the args object after using `$1` the first time, it's no longer there to be referenced a second time. I also don't like splicing because it breaks the relationship between argument index in the command string, and argument index in the mutated `args` array.